crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -131,46 +131,78 @@ impl CommandQueue {
         let bytes_range = self.bytes.as_mut_ptr_range();
 
         // Pointer that will iterate over the entries of the buffer.
-        let mut cursor = bytes_range.start;
+        let cursor = bytes_range.start;
+
+        let end = bytes_range.end;
 
         // Reset the buffer, so it can be reused after this function ends.
         // In the loop below, ownership of each command will be transferred into user code.
-        // SAFETY: `set_len(0)` is always valid.
-        unsafe { self.bytes.set_len(0) };
-
-        while cursor < bytes_range.end {
-            // SAFETY: The cursor is either at the start of the buffer, or just after the previous command.
-            // Since we know that the cursor is in bounds, it must point to the start of a new command.
-            let meta = unsafe { cursor.cast::<CommandMeta>().read_unaligned() };
-            // Advance to the bytes just after `meta`, which represent a type-erased command.
-            // SAFETY: For most types of `Command`, the pointer immediately following the metadata
-            // is guaranteed to be in bounds. If the command is a zero-sized type (ZST), then the cursor
-            // might be 1 byte past the end of the buffer, which is safe.
-            cursor = unsafe { cursor.add(std::mem::size_of::<CommandMeta>()) };
-            // Construct an owned pointer to the command.
-            // SAFETY: It is safe to transfer ownership out of `self.bytes`, since the call to `set_len(0)` above
-            // guarantees that nothing stored in the buffer will get observed after this function ends.
-            // `cmd` points to a valid address of a stored command, so it must be non-null.
-            let cmd = unsafe {
-                OwningPtr::<Unaligned>::new(std::ptr::NonNull::new_unchecked(cursor.cast()))
-            };
-            // SAFETY: The data underneath the cursor must correspond to the type erased in metadata,
-            // since they were stored next to each other by `.push()`.
-            // For ZSTs, the type doesn't matter as long as the pointer is non-null.
-            let size = unsafe { (meta.consume_command_and_get_size)(cmd, &mut world) };
-            // Advance the cursor past the command. For ZSTs, the cursor will not move.
-            // At this point, it will either point to the next `CommandMeta`,
-            // or the cursor will be out of bounds and the loop will end.
-            // SAFETY: The address just past the command is either within the buffer,
-            // or 1 byte past the end, so this addition will not overflow the pointer's allocation.
-            cursor = unsafe { cursor.add(size) };
+        let bytes = std::mem::take(&mut self.bytes);
```

Comment:
`take`ing the bytes buffer is unfortunate, since it means every command queue needs to be reallocated every frame. Can we hokey-pokey strats the buffer back into place after we're done applying?

Reply:
Since it's a `Vec` shouldn't `take` just be be re-assigning pointers?

Reply:
The issue is that after `apply` gets called, the buffer is empty, which means any commands that get pushed to it will trigger an allocation. On main, `CommandQueue`s usually only trigger allocations early on in an app's life cycle, as once the buffer grows large enough it can be reused for all future frames.

Reply:
Ah, true, we should be able to put the buffer back in place afterwards, the only struggle may be with juggling the other command buffers, I'll take a closer look when I do a pass.